pdm/models/readers.py

```diff
@@ -1,103 +1,105 @@
 import ast
 from configparser import ConfigParser
+from dataclasses import asdict, dataclass, field
 from pathlib import Path
-from typing import Any, Dict, Iterable, List, Optional, Tuple, Union
+from typing import Any, Dict, Iterable, List, Optional, Tuple
+
+
+@dataclass
+class Setup:
+    name: Optional[str] = None
+    version: Optional[str] = None
+    install_requires: List[str] = field(default_factory=list)
+    extras_require: Dict[str, List[str]] = field(default_factory=dict)
+    python_requires: Optional[str] = None
+
+    def update(self, other: "Setup"):
+        if other.name:
+            self.name = other.name
+        if other.version:
+            self.version = other.version
+        if other.install_requires:
+            self.install_requires = other.install_requires
+        if other.extras_require:
+            self.extras_require = other.extras_require
+        if other.python_requires:
+            self.python_requires = other.python_requires
+
+    def as_dict(self) -> Dict[str, Any]:
+        return asdict(self)
+
+    @classmethod
+    def from_directory(cls, dir: Path) -> "Setup":
+        return _SetupReader.read_from_directory(dir)
```

Comment:
Why prefer exposing `Setup` instead of `SetupReader`? You know, it may involve changing the reference in other modules as well.

Reply:
because i'm now exposing a strongly-typed 'Setup' class instead of a stringly-typed dictionary.

I've updated both external modules that reference this

Reply:
these changes bring the number of mypy errors down from 44 to 20. Still a few to go though! - https://github.com/frostming/pdm/pull/321#issuecomment-803589924

in the previous implementation the 'setup' is stored in a `Dict[str, Any]`, which means type inference doesn't work in any function that indexes any of the fields.

Now I could still have imported the `SetupReader` in other modules instead of `Setup`, but i think this is a neater abstraction. the `from_directory` static method on `Setup` is effectively a constructor, and the calling code doesn't need to know anything about the `SetupReader`, which is now effectively just a helper class